### DISCUSSION TITLE
Prevent the GC from compacting the heap

### DISCUSSION THREAD

**Discussion by m-carrasco** [dotnet/runtime#84625] [Q&A] [answered]
Hi,

I am interested in configuring the GC so that allocated objects are not moved around memory once created. I want them to live in the same address throughout their lives span. However, I want them swept once they are no longer used (whenever the GC thinks it is best). I think that in GC jargon, this translates to disabling compacting.  I'm happy by just achieving this for applications targeting `net7.0+` in x86-64/Linux.

After researching a bit, I think there are two possible ways. I would like to know if you think they would work and if you have any other advice. My intent is not to run this in production, and I understand that this is possibly not how the runtime should ever be configured.  

_First approach:_

AFAIK, the Large Object Heap is never compacted, which is what I want. I could adjust the [Large Object Heap Threshold](https://learn.microsoft.com/en-us/dotnet/core/runtime-config/garbage-collector#large-object-heap-threshold) to 1 or 0 bytes, so all objects are allocated in this way. This would also require setting [`GCLargeObjectHeapCompactionMode.Default`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.gclargeobjectheapcompactionmode?view=net-7.0#fields) at startup. 

_Second approach:_

I read [here](https://stackoverflow.com/questions/46574407/unitys-garbage-collector-why-non-generational-and-non-compacting) that one of Mono's GC does not compact the memory. Also, this [runtime's option](https://learn.microsoft.com/en-us/dotnet/core/runtime-config/garbage-collector#standalone-gc) allows changing the GC implementation providing it in a DLL. I am unsure if the Boehm GC could be set that way for applications targeting net7.0 in x86-64/Linux.

Any thoughts on which one could be easier to get working, or any other advice could be really helpful.

Thanks for your time!
Manuel.

**Comment by jkotas** [comment, marked as answer]
The easiest way to do this is to modify the place that decides whether the compact, force it to never compact and build your own GC with this modification. Look for `decide_on_compacting` in gc.cpp.

> First approach

The LOH threshold can be only set higher than the default. It cannot be set lower.

> Second approach

Replacing CoreCLR GC with Boehm GC would be a major project (months of work to even build a working prototype).

**Reply by m-carrasco** [reply]
Hi @jkotas :wave:

> The LOH threshold can be only set higher than the default. It cannot be set lower.

I overlooked the fact that the LOH can only be increased. As you said, this is written in the docs. When I tested the env variable, it did not report any error: `DOTNET_GCLOHThreshold=1 dotnet app.dll`.

> The easiest way to do this is to modify the place that decides whether the compact, force it to never compact and build your own GC with this modification. Look for decide_on_compacting in gc.cpp.

Following your advice, I will try manually disabling compacting. I would like to have a safeguard to ensure I did not forget to change any other place in the code. Do you know if this is possible? If I get it right, with these changes, I should only have objects in gen0 and the LOH. I am implicitly assuming that compacting is the only reason that may move an (alive) object in memory by the GC. Does this sound correct to you?

Thanks for your answer and insight! It is really helpful.

**Reply by jkotas** [reply]
> I should only have objects in gen0 and the LOH. I am implicitly assuming that compacting is the only reason that may move an (alive) object in memory by the GC

The objects can move between generations without moving in memory. The space assigned to each generation is not fixed. For example, when the Gen0 GC decides that it is not necessary to compact the heap, it will simply declare the space that used be Gen0 to be Gen1.
